### PR TITLE
YAML support in arg-parser

### PR DESCRIPTION
**Pull request by rahulbshrestha**
This PR allows yaml files to be read as config through the arg-parser.

Related issue: https://github.com/marrlab/DomainLab/issues/72

If a command line argument looks like this:

`python main_out.py --te_d=caltech --tpath=examples/tasks/task_vlcs.py --debug --bs=2 --aname=diva --gamma_y=7e5 --gamma_d=1e5 --nname=alexnet --nname_dom=conv_bn_pool_2
`

A yaml file for it would be:
```
---  # yaml document beginning

te_d: caltech
tpath: examples/tasks/task_vlcs.py
bs: 2
aname: diva
debug: True
gamma_y: 700000.0 #7e5?
gamma_d: 100000.0 #1e5?
nname: alexnet
nname_dom: conv_bn_pool_2

```
And to run this from the command line, the user would need to do:

`python main_out.py --config="tests/data/test.yaml"`

### CODE REVIEW

**Comment by rahulbshrestha** [comment]
TODOs:
[] Include yaml in dependencies to fix checks
[] add test
[] check how to read scientific values like "1e5" in yaml (is converted to 10000.0 when entered from command line)

**Comment by rahulbshrestha** [comment]
@smilesun 

(1) Is a test necessary for this? If yes, what would you recommend to test? 
(2) There is an issue with how scientific values are read from the yaml files. Entering --gamma_d=1e5 from the command line means it is read as 10000.0. But, it doesn't work the same way from the yaml file. Therefore the user will have to write it as "gamma_d: 10000.0" in the yaml file. Should I handle this case such that the yaml file also supports scientific values?
(3) Why is the check for "Upload coverage to Codecov" in CI/Run tests failing?

**Comment by smilesun** [comment]
> @smilesun
> 
> (1) Is a test necessary for this? If yes, what would you recommend to test? (2) There is an issue with how scientific values are read from the yaml files. Entering --gamma_d=1e5 from the command line means it is read as 10000.0. But, it doesn't work the same way from the yaml file. Therefore the user will have to write it as "gamma_d: 10000.0" in the yaml file. Should I handle this case such that the yaml file also supports scientific values? (3) Why is the check for "Upload coverage to Codecov" in CI/Run tests failing?
great job!
1) you mean test? definitely we need to test 95% of the lines you wrote
2) I think we can leave it for the first version, it is not that important
3) because the lines you wrote was not tested, the coverage should be 95% or more to pass the CI

File: domainlab/arg_parser.py [hunk 1]
```diff
@@ -148,8 +153,21 @@ def parse_cmd_args():
...
     args = parser.parse_args()
+    if args.config_file:
+
+        data = yaml.safe_load(args.config_file)
+        delattr(args, 'config_file')
+        arg_dict = args.__dict__
+
+        for key, value in data.items():
+            if isinstance(value, list):
+                arg_dict[key].extend(value)
```
Comment (smilesun) [inline domainlab/arg_parser.py:164]: do we need a warning here that if there exist some items in the list yaml file inputs will further expand the list?  can it happen that the user provided something which already exist in yaml file?

**Review by smilesun — CHANGES_REQUESTED** [review]
if the yaml file contains fields that does not exsit in argparser, we shall throw an exception?

File: domainlab/arg_parser.py (hunk 1 above)
Comment (rahulbshrestha) [reply domainlab/arg_parser.py:164]: I added a check to (1) throw an exception if yaml file contains fields not in argparser.

Comment (rahulbshrestha) [reply domainlab/arg_parser.py:168]: I added a check to (1) throw an exception if yaml file contains fields not in argparser.

> do we need a warning here that if there exist some items in the list yaml file inputs will further expand the list? can it happen that the user provided something which already exist in yaml file?

I'm not sure about this. I think users would either want to use a yaml file or just command-line arguments. I don't see a reason as to why one would choose both. In the current setup, the parameters in the yaml file would overwrite any command-line arguments if they are both the same.

**Review by smilesun — APPROVED** [review]
Good job Rahul!

**Pull request merged**